### PR TITLE
W6 -D1 -Lam Bui -Unit Test

### PR DESCRIPTION
**Pull request by at-lambui**
Task #62
Anh @at-trungnguyen ,em đã làm xong yêu cầu của anh.
Nhờ anh review giúp em.
Em cảm ơn.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -75,8 +75,10 @@ dependencies {
     implementation 'pub.devrel:easypermissions:0.2.0'
     implementation 'com.github.bumptech.glide:glide:4.7.1'
     annotationProcessor 'com.github.bumptech.glide:compiler:4.7.1'
-    apply from: "../ci-java.gradle"
     implementation 'com.squareup.retrofit2:retrofit:2.4.0'
     implementation 'com.squareup.retrofit2:converter-gson:2.2.0'
     implementation 'com.google.code.gson:gson:2.8.5'
+    testImplementation 'org.mockito:mockito-core:2.7.22'
+    androidTestImplementation 'org.mockito:mockito-android:2.7.22'
```
Comment (at-trungnguyen) [inline app/build.gradle]: remove line 82, it use for UI test.

File: app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java [hunk 2]
```diff
@@ -3,39 +3,67 @@
...
 import asiantech.internship.summer.viewpager.PagerActivity;
 
-public class MenuActivity extends AppCompatActivity {
+public class MenuActivity extends AppCompatActivity implements View.OnClickListener {
+    Button mBtnViewAndViewGroup;
+    Button mBtnFragmentAndListener;
+    Button mBtnRecyclerView;
+    Button mBtnViewPager;
+    Button mBtnUnitTest;
     @Override
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java]: break 1 line

File: app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java [hunk 3]
```diff
@@ -3,39 +3,67 @@
...
-        btnViewAndViewGroup.setOnClickListener(view -> {
-            Intent intent = new Intent(MenuActivity.this, ViewAndViewGruopActivity.class);
-            startActivity(intent);
-        });
-        Button btnViewpager = findViewById(R.id.btnViewPager);
-        btnViewpager.setOnClickListener(view -> {
-            Intent intentViewpager = new Intent(MenuActivity.this, PagerActivity.class);
-            startActivity(intentViewpager);
-        });
+        initViews();
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java]: good 👍

File: app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java [hunk 4]
```diff
@@ -3,39 +3,67 @@
...
+        addListeners();
+    }
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java]: enter

File: app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java [hunk 5]
```diff
@@ -3,39 +3,67 @@
...
+    private void initViews() {
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java]: enter

File: app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java [hunk 6]
```diff
@@ -3,39 +3,67 @@
...
+        mBtnViewAndViewGroup = findViewById(R.id.btnViewAndViewGroup);
+        mBtnFragmentAndListener = findViewById(R.id.btnExerciseFragment);
+        mBtnRecyclerView = findViewById(R.id.btnRecyclerview);
+        mBtnViewPager = findViewById(R.id.btnViewPager);
+        mBtnUnitTest = findViewById(R.id.btnUnitTest);
+    }
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java]: enter

File: app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java [hunk 7]
```diff
@@ -3,39 +3,67 @@
...
+    private void addListeners() {
+        mBtnViewAndViewGroup.setOnClickListener(this);
+        mBtnFragmentAndListener.setOnClickListener(this);
+        mBtnRecyclerView.setOnClickListener(this);
+        mBtnViewPager.setOnClickListener(this);
+        mBtnUnitTest.setOnClickListener(this);
+    }
+    @Override
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/menu/MenuActivity.java]: enter, method need space with other.

File: app/src/at-lambui/java/asiantech/internship/summer/unittest/ValidatorInit.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
+package asiantech.internship.summer.unittest;
+
+import java.util.regex.Pattern;
+
+import asiantech.internship.summer.R;
+import asiantech.internship.summer.unittest.model.User;
+
+public class ValidatorInit {
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/unittest/ValidatorInit.java]: `public final class`
define `private` contructor

File: app/src/at-lambui/java/asiantech/internship/summer/unittest/ValidatorInit.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            return R.string.error_username_have_not_at_least_two_uppercase_nonconsecutive;
+        }
+        return R.string.username_result;
+    }
+    /**
+     * @param password login
+     * @return password true
+     */
+    public static int validatePassword(String password) {
+
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/unittest/ValidatorInit.java]: why space here?

File: app/src/at-lambui/res/layout/activity_menu.xml [hunk 10]
```diff
@@ -41,4 +41,13 @@
...
 
+    <Button
+        android:id="@+id/btnUnitTest"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="10dp"
+        android:background="@color/colorPrimaryBlack"
+        android:text="@string/menu_activity_unit_test_title"
+        android:textSize="@dimen/font_normal" />
+
```
Comment (at-trungnguyen) [inline app/src/at-lambui/res/layout/activity_menu.xml]: delete this line 52

File: app/src/at-lambui/res/layout/activity_unit_test.xml [hunk 11]
```diff
@@ -0,0 +1,98 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (at-trungnguyen) [inline app/src/at-lambui/res/layout/activity_unit_test.xml]: format this file

File: app/src/test/java/lambui/test/PasswordTest.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        assertEquals(ValidatorInit.validatePassword(mUser.getPassWord()), R.string.error_password_have_not_enough_8_character_char);
+    }
+
+    @Test
+    public void passWordHaveCharacterRepeatMore() {
+        Mockito.doReturn("LamBui1233Tk").when(mUser).getPassWord();
+        assertEquals(ValidatorInit.validatePassword(mUser.getPassWord()), R.string.error_password_have_character_Which_Repeated);
+    }
+
+    //it nhat 2 ki tu so
```
Comment (at-trungnguyen) [inline app/src/test/java/lambui/test/PasswordTest.java]: no need comment, if need please use English.

File: app/src/at-lambui/java/asiantech/internship/summer/unittest/ValidatorInit.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        }
+        if (!PassWord_Not_End_With_Character_Special_0r_Digital.matcher(password).find()) {
+            return R.string.error_password_have_not_end_with_digital_or_char;
+        }
+        if (!PassWord_Have_At_Least_Two_Character_Digital_Or_Special.matcher(password).find()) {
+            return R.string.error_password_have_not_at_least_two_character_digital;
+        }
+
+        return R.string.password_result;
+    }
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/unittest/ValidatorInit.java]: enter

File: app/src/at-lambui/java/asiantech/internship/summer/unittest/ValidatorInit.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    public static int checkUserName(User user) {
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/unittest/ValidatorInit.java]: add `java doc`

File: app/src/at-lambui/java/asiantech/internship/summer/unittest/ValidatorInit.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+            Pattern.compile("^.*[a-zA-Z]$");
+    private static final Pattern PassWord_Have_At_least_3_Char_NonConsecutive =
+            Pattern.compile("^.*[A-Z].+[A-Z].+[A-Z]");
+    private static final Pattern PassWord_Have_At_Least_Two_Character_Digital_Or_Special =
+            Pattern.compile("^.*[^a-zA-Z].*[^a-zA-Z]");
+    /**
+     * @param username login
+     * @return username true
+     */
+
```
Comment (at-trungnguyen) [inline app/src/at-lambui/java/asiantech/internship/summer/unittest/ValidatorInit.java]: why space here.

File: app/src/at-lambui/res/drawable/custom_edit_text.xml [hunk 16]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (at-trungnguyen) [inline app/src/at-lambui/res/drawable/custom_edit_text.xml]: format this file

File: app/src/test/java/lambui/test/PasswordTest.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
+package lambui.test;
```
Comment (at-trungnguyen) [inline app/src/test/java/lambui/test/PasswordTest.java]: wrong package, fail CI. Please rename to `testAt-lambui`

File: app/src/test/java/lambui/test/UsernameTest.java [hunk 18]
```diff
@@ -0,0 +1,60 @@
+package lambui.test;
```
Comment (at-trungnguyen) [inline app/src/test/java/lambui/test/UsernameTest.java]: wrong package

File: app/src/testAt-lambui/java/asiantech/internship/summer/PasswordTest.java [hunk 19]
```diff
@@ -0,0 +1,77 @@
+package asiantech.internship.summer;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.mockito.Spy;
+
+import static junit.framework.Assert.*;
+
+import asiantech.internship.summer.R;
```
Comment (at-dev-support-01) [inline app/src/testAt-lambui/java/asiantech/internship/summer/PasswordTest.java]: WARNING: [ImportFromSamePackage] No need to import a type that lives in the same package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#ImportFromSamePackage

File: app/src/testAt-lambui/java/asiantech/internship/summer/PasswordTest.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Test
+    public void passWordHaveCharacterRepeatMore() {
+        Mockito.doReturn("LamBui1233Tk").when(mUser).getPassWord();
+        assertEquals(ValidatorInit.validatePassword(mUser.getPassWord()), R.string.error_password_have_character_Which_Repeated);
+    }
+
+    // at least two character digital
+    @Test
+    public void PassWordHaveAtLeastTwoCharacterDigitalOrSpecial() {
```
Comment (at-dev-support-01) [inline app/src/testAt-lambui/java/asiantech/internship/summer/PasswordTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-lambui/java/asiantech/internship/summer/UsernameTest.java [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+import org.junit.Before;
+
+import static junit.framework.Assert.*;
+
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+import asiantech.internship.summer.R;
```
Comment (at-dev-support-01) [inline app/src/testAt-lambui/java/asiantech/internship/summer/UsernameTest.java]: WARNING: [ImportFromSamePackage] No need to import a type that lives in the same package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#ImportFromSamePackage

File: app/src/testAt-lambui/java/asiantech/internship/summer/UsernameTest.java [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Before
+    public void init() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void usernameHaveLengthlessThan7char() {
+        Mockito.when(mUser.getUserName()).thenReturn("abac1");
+        Assert.assertEquals(ValidatorInit.validateUserName(mUser.getUserName()), R.string.error_username_length);
```
Comment (at-dev-support-01) [inline app/src/testAt-lambui/java/asiantech/internship/summer/UsernameTest.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'Assert.assertEquals' due to existing static import 'junit.framework.Assert.*'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/test/java/lambui/test/PasswordTest.java (hunk 17 above)
Comment (at-lambui) [reply app/src/test/java/lambui/test/PasswordTest.java]: Ok , I fixed fail CI, many thanks.

File: app/src/testAt-lambui/java/asiantech/internship/summer/UsernameTest.java [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
```
Comment (at-dev-support-01) [inline app/src/testAt-lambui/java/asiantech/internship/summer/UsernameTest.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'Assert.assertEquals' due to existing static import 'junit.framework.Assert.assertEquals'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

**Pull request merged**